### DISCUSSION TITLE
prevent duplicate vendor

### DISCUSSION THREAD

**Discussion by rino1** [composer/composer#11403] [General]
I created two packages.
Here are the related composer.json files

````
{
    "name": "my/sub1",
    "version": "1",
    "require": {
       "my/sub2": "^1.0"
    }
}

{
    "name": "my/sub2",
    "version": "1"
}
````

Note that the my/sub1 package requires the my/sub2 package.

We download

````
$ composer require my/sub1
[CUT]
````

Let's see what happened

````
$ find
[ONLY SUMMARY]
./vendor/my/sub1
./vendor/my/sub1/vendor/my/sub2
./vendor/my/sub2
````

But I would have liked to be downloaded only the ./vendor/my/sub1 and ./vendor/my/sub1/vendor/my/sub2 directory, and not also ./vendor/my/sub2
It's possible?

**Comment by stof** [comment]
`/vendor/my/sub1/vendor/my/sub2` should not exist, unless you committed your vendor folder inside the `my/sub1` package (which should not be done).

Composer will never create a `vendor` folder nested inside an installed vendor package.

**Reply by rino1** [reply]
Let's do something more detailed
````
$ ssh myname@remote-repo

remote-repo:~# cd /my/sub1

remote-repo:/my/sub1# composer update
[CUT]
remote-repo:/my/sub1# test -e vendor/my/sub2 && echo "DIRECTORY FOUND" && exit
DIRECTORY FOUND

$ composer require my/sub1
[CUT]

$ find
[ONLY SUMMARY]
./vendor/my/sub1
./vendor/my/sub1/vendor/my/sub2
./vendor/my/sub2
````

**Reply by stof** [reply]
with those anonymised packages with no idea where they come from, there is nothing I can do to help you understand why this `/vendor/my/sub1/vendor` folder exists

**Reply by rino1** [reply]
````
+ rm -rf one/composer.lock one/vendor root/composer.json root/composer.lock root/vendor
+ tree -a
.
├── .config -> ../.config/
├── one
│   ├── composer.json
│   └── script
├── response
├── root
└── two
    ├── composer.json
    └── script

4 directories, 5 files
+ cat .config/composer/config.json
{
    "repositories": {
        "one": {
            "type": "path",
            "url": "$PWD/one"
        },
        "two": {
            "type": "path",
            "url": "$PWD/two"
        }
    }
}
+ cat one/composer.json
{
    "name": "package/one",
    "version": "1",
    "require": {
       "package/two": "^1.0"
    }
}
+ cat two/composer.json
{
    "name": "package/two",
    "version": 1
}
+ cat one/script
#!/bin/bash -eu
echo $0: Hello world!
# do something...
sub_script=$(/usr/bin/dirname "$0")/vendor/package/two/script
$sub_script && echo OK: $sub_script called
# do something...
+ cat two/script
#!/bin/bash -eu
echo $0: Hello world!
# do something...
+ one/script
one/script: Hello world!
one/script: line 5: one/vendor/package/two/script: No such file or directory
+ composer update -d one
Loading composer repositories with package information
Updating dependencies
Lock file operations: 1 install, 0 updates, 0 removals
  - Locking package/two (1)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Installing package/two (1): Symlinking from /var/tmp/duplicate-vendor/public/two
Generating autoload files
+ one/script
one/script: Hello world!
one/vendor/package/two/script: Hello world!
OK: one/vendor/package/two/script called
+ composer require -d root package/one
Using version ^1.0 for package/one
./composer.json has been created
Running composer update package/one
Loading composer repositories with package information
Updating dependencies
Lock file operations: 2 installs, 0 updates, 0 removals
  - Locking package/one (1)
  - Locking package/two (1)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 2 installs, 0 updates, 0 removals
  - Installing package/two (1): Symlinking from /var/tmp/duplicate-vendor/public/two
  - Installing package/one (1): Symlinking from /var/tmp/duplicate-vendor/public/one
Generating autoload files
````

Now let's pay close attention

````
+ cat response
#!/bin/bash -u
rootVendorDir=root/$1
oneVendorDir=one/$1
if diff -rq $rootVendorDir $oneVendorDir; then
    echo $rootVendorDir it a duplicate of $oneVendorDir
    if ! grep -r $1 root && grep -r $1 one; then
        echo "Only 'one' package use $oneVendorDir. Therefore downloading of $rootVendorDir should be avoided."
    fi
fi
````

Final verdict

````
+ ./response vendor/package/two
root/vendor/package/two it a duplicate of one/vendor/package/two
one/script:sub_script=$(/usr/bin/dirname "$0")/vendor/package/two/script
Only 'one' package use one/vendor/package/two. Therefore downloading of root/vendor/package/two should be avoided.
````

**Reply by stof** [reply]
OK, this is because you use a `path` repository, which creates a symlink, while you also installed dependencies inside that path by using it as a root package. That's one `one/vendor/package/two` exists.
But it your main project, this `one/vendor/package/two` won't be used as composer won't be aware about it and won't register any autoloading for it.

**Reply by rino1** [reply]
FALSE!!!

I had forgotten...

```
+ cd root
```

I created the file my-script


```
+ cat my-script
#!/bin/bash
$(/usr/bin/dirname "$0")/vendor/package/one/script

+ ./my-script
./vendor/package/one/script: Hello world!
./vendor/package/one/vendor/package/two/script: Hello world!
OK: ./vendor/package/one/vendor/package/two/script called

+ !! | grep one/vendor/package/two && echo one/vendor/package/two FOUND
./vendor/package/one/vendor/package/two/script: Hello world!
OK: ./vendor/package/one/vendor/package/two/script called
one/vendor/package/two FOUND
```

**Reply by rino1** [reply]
After this umpteenth clarification, however, I expected a courtesy response (at least one opinion) on the final verdict: https://github.com/composer/composer/discussions/11403#discussioncomment-5465545 (last part)

**Comment by rino1** [comment]
I resume the discussion for some clarification

> `/vendor/my/sub1/vendor/my/sub2` should not exist, unless you committed your vendor folder inside the `my/sub1` package (which should not be done).

But if I have unit tests to be performed, I have to give the command

`composer require --dev phpunit/phpunit && vendor/bin/phpunit`

which will inevitably be to create the vendor directory.

So what do you do?

**Comment by rino1** [comment]
Nothing?

**Reply by mbabker** [reply]
This is just how Composer works, I suggest taking a read over https://getcomposer.org/doc/05-repositories.md#path for more information.

When Composer installs a package from a path repository and it makes a symlink, then whatever you've got loaded in that symlinked path is going to be present in the parent project.  The presence of a `vendor` folder in one of those symlinked projects shouldn't be an issue unless something is loading the autoloader from that nested vendor (so the main application ends up loading `vendor/autoload.php` and `vendor/anon/package1/vendor/autoload.php` as an example), it'll be annoying in an IDE if it shows multiple class declarations but that's really the extent of the damage that should be able to do.

**Comment by rino1** [comment]
So the ideal would be to avoid having the vendor/anon/package1/vendor directory, avoiding to run


```
composer require -d vendor/anon/package1 --dev phpunit/phpunit
```

But if I have unit tests into anon/package1 to be run, I have to give the command

```
vendor/anon/package1/vendor/bin/phpunit
```

This forces me to download the phpunit/phpunit composer package, otherwise would answer me

```
vendor/anon/package1/vendor/bin/phpunit: not found
```

That's what I had already said here https://github.com/composer/composer/discussions/11403#discussioncomment-9115041

So what do you do?

Unfortunately @stof  no longer gave any answer

**Reply by stof** [reply]
Composer is meant to install all dependencies in a **flat** `vendor` folder. It is not expected to handle nested vendor folders. This is related to the fact that PHP class names must be unique per PHP process, so having 2 versions of the package installed in different locations will create bugs (especially if they are _different_ versions).

Using the same path both as a root package to work inside it and as a path repository to be symlinked as a dependency elsewhere is a weird setup, for which you will indeed see duplicate class definitions in your IDE, unless you configure your IDE to ignore the vendor folders of your path repository (as answered by @mbabker a few months ago, just before your message). But as long as your main application never loads the autoload files from such nested folders (which composer will never do automatically, unless your `path` packages have an autoload config that includes everything in it and not just an `src` folder), the nested vendor folders caused by that weird setup should not have an impact as the main app will never use them.

Side note: the discussion would be a lot more understandable if you were actually writing a reply when replying to a previous message instead of adding a new unrelated comment).
Another side note: replying on such discussions on the Composer repo is something I do on my free time (and without being an actual maintainer of composer, only a contributor knowing the codebase very well due to contributing since years).

**Reply by rino1** [reply]
> Using the same path both as a root package to work inside it and as a path repository to be symlinked as a dependency elsewhere is a weird setup

So what approach would you take when dealing only with local repositories (no git)?
For example, keeping two separate directories (xxx and xxx.dev) to synchronize periodically `rsync -ai --delete --exclude vendor xxx.dev xxx`?
I don't know...

**Reply by stof** [reply]
When you say "no git", you mean that you don't even use git locally to track changes in what you work on ? Then I don't know what I would do in such case as I haven't worked on non-versioned-controlled code since I discovered git (I would probably run `git init`, actually).
If this is for the case of a mono-repo containing both the package and the app using it, maybe installing composer dependencies at the monorepo level during development could be a way to avoid the issue

It is also possible with Composer to use a git repository pointing to a local path (this does not provide the same experience than a `path` repository as it would require running a `composer update` to update to a new commit of the referenced repository, as it is not a symlink, but a solution based on `rsync` also requires running a command to get changes synced). This is done by making it point to the `.git` folder inside your local clone.

**Reply by rino1** [reply]
> When you say "no git", you mean that you don't even use git locally to track changes in what you work on ?

YES

> Then I don't know what I would do in such case as I haven't worked on non-versioned-controlled code since I discovered git (I would probably run `git init`, actually). If this is for the case of a mono-repo containing both the package and the app using it, maybe installing composer dependencies at the monorepo level during development could be a way to avoid the issue

It is not a mono-repo
 
> It is also possible with Composer to use a git repository pointing to a local path (this does not provide the same experience than a `path` repository as it would require running a `composer update` to update to a new commit of the referenced repository, as it is not a symlink, but a solution based on `rsync` also requires running a command to get changes synced). This is done by making it point to the `.git` folder inside your local clone.

I don't know how to do such a thing.
I have never used software to control the version.
I should document me ...

An example on the fly?

**Comment by rino1** [comment]
Since I didn't get any answers, I tried to do some intuitive tests

```
+ tree
├── composer1.json
├── composer2.json
└── package
    └── composer.json

+ cat ./package/composer.json
{"name": "my/package"}

+ cat ./composer1.json
{
    "repositories": {
        "package": {
            "type": "path",
            "url": "$PWD/package"
        }
    }
}

+ cat ./composer2.json
{
    "repositories": {
        "package": {
            "type": "vcs",
            "url": "$PWD/package"
        }
    }
}

$ COMPOSER=composer1.json composer -a show my/package
name     : my/package
descrip. : 
keywords : 
versions : dev-main
blabla...
```

OK

```
$ COMPOSER=composer2.json composer -a show my/package
In VcsRepository.php line 192:
  No driver found to handle VCS repository /var/tmp/vcs/package  
```

:(

**Comment by rino1** [comment]
UPPPPPPPPPPPPPP

**Reply by mbabker** [reply]
You're not going to find an answer that satisfies your workflow on this board.  Folks have tried helping you but to be completely honest the way you're going about working on these things means you're going to have to accept the tradeoffs of those decisions.

You're developing a package standalone from an application without using version control, then you're symlinking a directory into an application which also doesn't use version control, and expecting Composer to change the behavior of how an operating system's symlink works to suit your workflow.  If you don't want the embedded vendor folder, the symlink workflow isn't going to work.  If your issue is code is being loaded from that embedded vendor folder, then your application or package is doing something wrong because nothing should be arbitrarily loading every nested vendor folder.  If your issue is your IDE is showing duplicate definitions, silence those warnings in your IDE.

**Reply by rino1** [reply]
> You're developing a package standalone from an application without using version control,
Let me understand, should I install a locally git server?

I do not speak English, and if the concepts do not come to me clearly explained, you do not derive us

**Reply by rino1** [reply]
> You're developing a package standalone from an application without using version control,


Let me understand, should I install a locally git server?

I do not speak English, and if the concepts do not come to me clearly explained, you do not derive us
